### ISSUE TITLE
Convert indentation from smart tabs to 4-space

### ISSUE THREAD

**Issue by mourner** [Leaflet/Leaflet#3870]
I've been a fan of smart tabs (tabs for indentation, spaces for alignment) and persevered with it through many years, but have to admit that I regret this choice now and 4 spaces provide a much less painful experience for both maintainers and contributors. I now use 4 spaces in all my other projects (smooth sailing), and it would be great to convert Leaflet as well eventually.

These are the big problems I see with smart tabs:

- they're very hard to enforce on the linting level; there are all sorts of edge cases and the approach is not used often enough to drive fixing related issues and bringing related features on projects like ESLint
- most contributors are unfamiliar with the approach so I have to write "please fix whitespace" comments in PRs over and over again; it will get easier with https://github.com/Leaflet/Leaflet/commit/301473ad4d83ec14254464e2f598e9827a904927 but that's still a hack
- all-spaces indentation is much better supported in editors unlike mixed tabs and spaces, which is more hassle to write
- smart tabs bring complexity to whitespace management; all-spaces are barebones simple

These are the issues that prevent me from doing a repo-wide conversion:

- it'll **bork all the open PRs** (authors will have to painfully rebase and solve lots of conflicts); does anyone know a solution to this? I know there's at least an option to ignore whitespace when showing diffs, but is there an easy way to merge code with different whitespace in git?
- it'll screw `git blame` so that you'll have to run it again on commits before the conversion, but I guess that's a necessary evil

Should we go ahead with this and deal with consequences, or just leave this alone and continue to deal with whitespace annoyances?

cc @IvanSanchez @yohanboniface @danzel @jfirebaugh

**Comment by yohanboniface** [comment]
I'd support this and think the 1.0 is the good opportunity to do so. Will be painful, but necessary for a long term view imho.

**Comment by IvanSanchez** [comment]
Apparently it is possible to rewrite history and replace all tabs with spaces without borking `blame`, by using https://git-scm.com/docs/git-filter-branch and a `sed` replacement.

**Comment by mourner** [comment]
@IvanSanchez nah, that's just too scary — it would change all commit refs and break everything for sure. Not an option.

> WARNING! The rewritten history will have different object names for all the objects and will not converge with the original branch. You will not be able to easily push and distribute the rewritten branch on top of the original branch. Please do not use this command if you do not know the full implications, and avoid using it anyway,

**Comment by danzel** [comment]
Breaking blame is gonna suck, but it is easy to work around when necessary.
Any PRs can be redone if they are worthwhile.

You could do the commit from another account just for this (leaflet-whitespace-rewrite) to make it super obvious in the history.

This could be the final commit for 1.0? :)

**Comment by yohanboniface** [comment]
About white spaces only modification, `git blame` doc says:

```
-w
    Ignore whitespace when comparing the parent’s version and the child’s to find where the lines came from.
```
Not sure this apply to Github blame pages, though.

**Comment by Alhadis** [comment]
What's so difficult to enforce? I've [never had an issue](https://github.com/Alhadis/Coding-Style#2-dont-use-tabs-for-alignment).

Keeping tabs and spaces separated by printing/non-whitespace characters is as easy to check as `/\t\x20/`.

**Comment by septatrix** [comment]
Regarding the drawback for blame there is [`git blame --ignore-rev/--ignore-revs-file`](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt) which is also [supported by github when given the correct filename](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)

**Comment by Falke-Design** [comment]
@mourner do you still see the need to switch from tabs to spaces?

I'm pretty fine with the tabs as identation and there are not much _"please fix whitespace"_ reviews anymore.

**Comment by Falke-Design** [comment]
> Regarding "but somewhere we discussed that it can destroy the git history" from [#9652 (comment)](https://github.com/Leaflet/Leaflet/pull/9652#discussion_r2051789597):
> 
> The history is kept, and git will create new objects in `.git/objects`. It is a matter of configuring the diffing/blaming, e.g. https://github.com/Leaflet/Leaflet/pull/9664/files?w=1 `w=1` hides white space changes. And `git blame` may be configured to ignore certain commit hashes, see https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt
> 
_Originally posted by @simon04 in https://github.com/Leaflet/Leaflet/discussions/7729#discussioncomment-12892330_

**Comment by mourner** [comment]
@Falke-Design I'm fine either way — what do others think? In my other projects, I switched to spaces — much more common across the community and easier to keep clean because it doesn't mix tabs and spaces, but I also realize the hassle of a project-wide conversion. I realize you can view changes with white space ignored or configure git blame, but still, the default GitHub UI experience suffers and there seems to be no way around it.
